### ISSUE TITLE
Maintainers: please don't push directly to master/release branches

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#118661]
It seems like the community is close to a consensus for **strongly discouraging direct pushes** to master and release branches.

Mainly because nixpkgs is a huge project and we've hit the critical mass where things break too often and changes benefit from a review. We can not rely on discipline any longer.

There are exceptions that are allowed for the time being:

- automatic updates: those should transition to https://cli.github.com/manual/gh_pr_create or plain github actions
- release managers
- time-sensitive changes (evaluation fix, security fix, etc)
- trivial changes (typos)

For trivial changes like typos, you can prepend `[skip ci]` to commit message.

Note that there was https://github.com/NixOS/rfcs/pull/79 that will get a successor to formalize this.

**Comment by 7c6f434c** [comment]
>Mainly because nixpkgs is a huge project and we've hit the critical mass where things break too often. We can not rely on discipline any longer.

Break = channel stall? I wonder how much average channel advance frequency is actually affected by broken direct pushes as opposed to everything else.

>There are exceptions that are allowed for the time being:

Is leaving out backports a typo?

>Note that there was https://github.com/NixOS/rfcs/pull/79 that will get a successor to formalize this.

Not sure how relevant one more failed RFC is to anything…

**Comment by domenkozar** [comment]
> Break = channel stall? I wonder how much average channel advance frequency is actually affected by broken direct pushes as opposed to everything else.

Anything that doesn't pass the CI (excluding builds for the moment), which breaks all subsequent commits and PRs. This has happened many times in the past and PRs give a chance for everyone to review things.

> Is leaving out backports a typo?

I don't understand.

**Comment by FRidh** [comment]
Looking back at the old RFC, my comments regarding staging-(next) were irrelevant; as long as pushes remain possible to staging(-next), it is fine. Now, I would suggest to block pushing to master, and discourage on all other branches. That way, backports will at least keep happening. But let's discuss that further in an RFC.

**Comment by andir** [comment]
> Looking back at the old RFC, my comments regarding staging-(next) were irrelevant; as long as pushes remain possible to staging(-next), it is fine. Now, I would suggest to block pushing to master, and discourage on all other branches. That way, backports will at least keep happening. But let's discuss that further in an RFC.

I still wonder where the idea of "just pushing a backport" is different from a change to master comes from. It might as well be a breaking change even if the commit applies. IMHO maintainers should also be involved in those and especially in stable updates. Those might each have a different impact from package to package.

**Comment by FRidh** [comment]
From a "does it break something" point of view, I agree, it is not different. Impact wise, however, it is different; there are far fewer PR's made against stable branches so the impact of a failing CI on other PR's is less big. On the other hand, it could block a stable channel from advancing. I guess its a matter of what odds do you accept.

**Comment by 7c6f434c** [comment]
>Anything that doesn't pass the CI (excluding builds for the moment), which breaks all subsequent commits and PRs. 

In this sense, it is a trade-off in terms of development annoyance, not a user-facing breakage. For all the twelve years I remember, there were some annoying breakages on master from time to time (actually now they matter less than initially when Hydra substituters were much more entire-channel-or-nothing)

>> Is leaving out backports a typo?
>I don't understand.

You start with saying «master and release branches». Recently I have been under impression that a non-release-manager committer doing a backport to a release branch via a direct push is considered acceptable, because backports are not done enough. (I don't have an understanding what is backport-worthy in the first place as I never use stable branches)

**Comment by domenkozar** [comment]
>  this sense, it is a trade-off in terms of development annoyance, not a user-facing breakage.

Yes, we're talking about users that contribute to Nixpkgs.

> You start with saying «master and release branches». Recently I have been under impression that a non-release-manager committer doing a backport to a release branch via a direct push is considered acceptable, because backports are not done enough. (I don't have an understanding what is backport-worthy in the first place as I never use stable branches)

It is acceptable but strongly discouraged. 

Note that we have no proof that going through PRs will mean there are fewer backports, even more so if we create automation (a shell script with a few lines calling github cli).

Note 2: Removing the power from people always triggers "oh I don't like that so I won't participate" response, but that's not a sufficient proof :)

Note 3: We could use something like https://github.com/tibdex/backport to make this even easier than it ever was.

**Comment by 7c6f434c** [comment]
>> You start with saying «master and release branches». Recently I have been under impression that a non-release-manager committer doing a backport to a release branch via a direct push is considered acceptable, because backports are not done enough. (I don't have an understanding what is backport-worthy in the first place as I never use stable branches)
>
>It is acceptable but strongly discouraged. 
>
>Note that we have no proof that going through PRs will mean there are fewer backports, even more so if we create automation (a shell script with a few lines calling github cli).
>
>Note 2: Removing the power from people always triggers "oh I don't like that so I won't participate" response, but that's not a sufficient proof :)

It's true I can't say that about myself — I do not do backports at all anyway.

>Note 3: We could use something like https://github.com/tibdex/backport to make this even easier than it ever was.

I sure support the part about enabling that…

**Comment by vcunat** [comment]
I've been still doing direct pushes to nixpkgs master occasionally, always in cases that:
- are unlikely to get any feedback (from previous experience), _and_
- don't change anything critical, _and_
- are very unlikely to break anything (or anything else, at least, e.g. trivial version+hash bumps)

I could do them through PRs, though it'd be some extra hassle that didn't seem worth it under those conditions. (They don't seem covered by the "trivial changes" bullet, as you formulate it.)

**Comment by bergkvist** [comment]
I'm not a member of the Nix organisation, but some potentially useful questions to ask here are:

- What is the probability that a push directly to master by a NixOS-member will introduce a new issue?

- How much lower is the probability that a pull request by a NixOS-member will introduce a new issue compared to a direct push?

- Does forcing everything through PRs motivate less atomic, and larger commits/changes?

- Enforcing PRs will likely reduce the output of the core NixOS members. By how much?


### A practical example
*(with made up numbers)*
Pushing directly to master: 10% chance of introducing breakage
Creating a PR: 3% chance of introducing breakage
Having to go through PRs instead of pushing directly halves the number of bugfixes that are shipped

With these made up numbers, what will lead to the fewest issues over time? Allowing direct pushes to master, or forcing everything to go through PRs?

---

It is not obvious that enforcing PRs for core maintainers will lead to a more "polished product". Rather, over time - it might not be able to keep up, and end up accumulating more issues.

Measuring what these numbers actually are might provide some useful insights.

**Comment by grahamc** [comment]
In my opinion a major part of the issue here is more than just reducing chances of introducing problems, but adding a chance for audit and review. In this day and age the number of projects where just pushing to the main branch is acceptable is quite low. I feel it is our responsibility as a distribution to take our role quite seriously and reduce the ways weird patches can be submitted without audit.

**Comment by bergkvist** [comment]
@grahamc Sure, auditing/reviewing everything sounds like a nice ideal, but this might be an issue:

> As has been already pointed out, the four eyes principle is hugely impractical for Nixpkgs, when we can't even get four eyes on mostsmall PRs as it is.

https://github.com/NixOS/rfcs/pull/79#issuecomment-714543485

**Comment by grahamc** [comment]
Yes I understand, however there is a big difference between no opportunity to review and a PR which was unreviewed.This is a step in the right direction.

**Comment by vcunat** [comment]
The current accumulated count of open vs. closed PRs doesn't seem that bad (2&ndash;3 %).

**Comment by grahamc** [comment]
> The current accumulated count of open vs. closed PRs doesn't seem that bad (2–3 %).

This makes no requirement on review, either. It simply gives the opportunity for review.

**Comment by vcunat** [comment]
If so, we might want to define how long an opportunity should be required.  We have at least one PR in nixpkgs that was self-merged within a couple seconds ;-)  (you might recall that from a release manager presentation on some NixCon)

**Comment by bergkvist** [comment]
Isn't it also possible to comment on individual commits? What is the difference between commenting on a merged PR and a commit? Two things I can think of:
- Might be harder to find/search for comments on commits in the GitHub UI
- Test suite doesn't run

**Comment by grahamc** [comment]
> If so, we might want to define how long an opportunity should be required. We have at least one PR in nixpkgs that was self-merged within a couple seconds ;-) (you might recall that from a release manager presentation on some NixCon)

Right. I'm actually not too worried about this, though. Simply having people use PRs and having ofborg check the evaluation is a net win. Reducing the direct push is a significant win. Removing the ability for almost everybody to directly push is another significant win. IMO creating more policy later is okay.

**Comment by bergkvist** [comment]
A complexity that arises from doing everything through PRs that are not immediately merged is bit rot (when the change is no longer valid/good due to other changes in the system), and the need for managing dependencies between PRs.

**Comment by grahamc** [comment]
@bergkvist, I feel you're here sea-lioning an issue that has no effect on you. Bitrot happens to all sorts of contributors. It is true. It impacts non-merging contributors more than anyone else, and people with merge access are the least impacted. Being asked to go through PRs is not likely to significantly hinder committers.

**Comment by bergkvist** [comment]
@grahamc You're right. Thanks for pointing it out. This is an issue which doesn't impact me directly - so I should also not have any impact on this issue. (Because I wouldn't personally feel the pain of having made a bad choice)

On a side-note, I really appreciate the work that you put into maintaining nixpkgs - and for making it the best package manager out there (in my opinion).

**Comment by domenkozar** [comment]
See https://botsin.space/@complainingaboutmastercommits for some statistics (I haven't verified if those numbers are correct).

**Comment by FRidh** [comment]
> See https://botsin.space/@complainingaboutmastercommits for some statistics (I haven't verified if those numbers are correct).

I am pretty sure those number are incorrect for the reason mentioned in https://github.com/NixOS/rfcs/pull/79#discussion_r508760845. Same author.

**Comment by matthiasbeyer** [comment]
>See ***@***.*** for some
>statistics (I haven't verified if those numbers are correct).

They are probably not, but IMO a good approximation.

Also,... Squash merges can break master, too, right?

**Comment by NeQuissimus** [comment]
Considering I have broken `master` on more than one occasion, I figured I'd chime in :)
I have finally convinced myself to use PRs for everything and from first-hand experience have multiple observations:
- Less breakage
- Availability of @ofborg for test evaluation is huge, I actually found myself adding tests on a regular basis
- There have actually been more people in the community giving feedback, reducing my workload (:pray:)

I think a few questions should be answered to make this process even better (and the RFC mentions these):
- Auto-merging for PRs that have tests
- I have a hard time with the distinction between `master`, `staging`, `staging-next` and equivalent stable branches. What goes where? How often are the branches merged into each other? This may just be my lack of insight... I think we need a way for an automatic label or a bot to ensure changes are sent to the correct target branches
- Is there a way we could ask @ofborg to take a PR and also evaluate against the stable branch(es)? Say I make a PR with kernel updates. If I could just `@ofborg stable 20.09` and it duplicates all its checks with the additional set running against 20.09, that would be fantastic. Otherwise the turnaround time would likely double before updates hit stable (Unless we do it entirely in parallel but then we cannot `git cherry-pick`, can we?)

**Comment by matthiasbeyer** [comment]
>  I think we need a way for an automatic label or a bot to ensure changes are sent to the correct target branches

Aren't the tags about how much gets rebuilt by a PR automatically added by a bot? Those could serve well for that purpose, I suppose?

**Comment by NeQuissimus** [comment]
Totally, but it currently just states the number of rebuilds. It would be easier if it also told me which branch to target (and maybe even blocks merging to the wrong branch)

**Comment by domenkozar** [comment]
> I think a few questions should be answered to make this process even better (and the RFC mentions these):
> 
> * Auto-merging for PRs that have tests

I don't think that's a good idea, automation should be used to make things easier, but not to play the role of review/approval. Tests already increase confidence when merging, so that helps a lot.

> * I have a hard time with the distinction between `master`, `staging`, `staging-next` and equivalent stable branches. What goes where? How often are the branches merged into each other? This may just be my lack of insight... I think we need a way for an automatic label or a bot to ensure changes are sent to the correct target branches

That's something yet to be well documented, but there's an RFC for staging/staging-next. For stable releases we backport bug fixes and new packages.

> * Is there a way we could ask @ofborg to take a PR and also evaluate against the stable branch(es)? Say I make a PR with kernel updates. If I could just `@ofborg stable 20.09` and it duplicates all its checks with the additional set running against 20.09, that would be fantastic. Otherwise the turnaround time would likely double before updates hit stable (Unless we do it entirely in parallel but then we cannot `git cherry-pick`, can we?)

It should be possible with https://github.com/tibdex/backport

**Comment by Ma27** [comment]
I don't mind switching to a fully PR-based workflow (though I also did direct pushes for mostly small packages bumps - so basically the same as @vcunat was doing).

However I have a few notes / questions:

> It should be possible with https://github.com/tibdex/backport

Doesn't this conflict with @andir's statement regarding backports requiring a review since these can also be breaking? Also, sorry for being *that* guy but I regularly see people criticizing the fact we're using GitHub / centralized infrastructure (and I kinda agree with that). So I'd like to raise the question whether we want to depend even more on GitHub's infrastructure, though I'm well-aware that this is an implementation-detail in this case.

> I still wonder where the idea of "just pushing a backport" is different from a change to master comes from. It might as well be a breaking change even if the commit applies. IMHO maintainers should also be involved in those and especially in stable updates. Those might each have a different impact from package to package.

Even though I mentioned that this conflicts with e.g. Domen's suggestion, I'm not sure I agree here:

When PRs are supposed to be backported, we have a `needs: port to stable`-label. Or in other words: (at least) I usually bring this topic up when a PR is open. So finding out whether it's fine to backport is already part of the "main" PR and that's why I actually consider a backport (i.e. a push to a release branch) part of the merge process.

Of course this doesn't apply to patches that are e.g. pretty huge or have non-trivial conflicts after a cherry-pick. Then a PR is obviously necessary.

> This has happened many times in the past and PRs give a chance for everyone to review things

Also I'd like to suggest to reflect our way of reviewing things:

We currently have 2.4k open PRs. I agree that it makes sense to audit changes, to check for potential breakage or bugs etc. But I don't think it helps (for newcomers and for keeping the queue under control) to leave nit-picky comments such as suggesting to write `callPackage foo/bar { }` rather than `callPackage foo/bar {}` or suggesting changes that are not related to the patch itself, but apply to (usually small) issues that happen to be in the same file.

Most of this can be automated by a tool and shouldn't require human effort. I personally think that focusing on things in changes that actually matter (well - from my PoV at least, feel free to disagree!) would make reviews more efficient and easier to justify (in contrast to "unreviewed" changes).

**Comment by vcunat** [comment]
> - I have a hard time with the distinction between `master`, `staging`, `staging-next` and equivalent stable branches.  What goes where?  [...]

I thought `staging*` was [documented](https://nixos.org/manual/nixpkgs/stable/#submitting-changes-staging-branch), but the freeze-RFC (0085) will unfortunately complicate the rules.  And I guess it could be more explicit about what's a "mass-rebuild" and mention the ofborg labels.

**Comment by domenkozar** [comment]
> I don't mind switching to a fully PR-based workflow (though I also did direct pushes for mostly small packages bumps - so basically the same as @vcunat was doing).
> 
> However I have a few notes / questions:
> 
> > It should be possible with https://github.com/tibdex/backport
> 
> Doesn't this conflict with @andir's statement regarding backports requiring a review since these can also be breaking? Also, sorry for being _that_ guy but I regularly see people criticizing the fact we're using GitHub / centralized infrastructure (and I kinda agree with that). So I'd like to raise the question whether we want to depend even more on GitHub's infrastructure, though I'm well-aware that this is an implementation-detail in this case.

Sorry, I didn't pay enough attention to see that it doesn't open a PR.

I've found https://github.com/tibdex/backport that does do what we expect (might need some tweaks, I haven't tested it).

About the GitHub "vendor lock-in": 

That's a trade-off between different resources, there's no perfect path here.

If we reuse GitHub actions, we get a lot of automation for free, if we write our own (by let's say extend ofborg) there's a lot of code to be written.

I'm not a fan of future-proofing by imposing requirements that might materialize one day. It increases the scope of work to be done. Maybe we'll stick with GitHub for another 10 years and then use something completely different, so a vendor-free solution is also useless at that point. Who knows?

If someone volunteers to take that risk, I'm up for it. But I wouldn't let the perfect be the enemy of the good here.

> > I still wonder where the idea of "just pushing a backport" is different from a change to master comes from. It might as well be a breaking change even if the commit applies. IMHO maintainers should also be involved in those and especially in stable updates. Those might each have a different impact from package to package.
> 
> Even though I mentioned that this conflicts with e.g. Domen's suggestion, I'm not sure I agree here:
> 
> When PRs are supposed to be backported, we have a `needs: port to stable`-label. Or in other words: (at least) I usually bring this topic up when a PR is open. So finding out whether it's fine to backport is already part of the "main" PR and that's why I actually consider a backport (i.e. a push to a release branch) part of the merge process.
> 
> Of course this doesn't apply to patches that are e.g. pretty huge or have non-trivial conflicts after a cherry-pick. Then a PR is obviously necessary.
> 
> > This has happened many times in the past and PRs give a chance for everyone to review things
> 
> Also I'd like to suggest to reflect our way of reviewing things:
> 
> We currently have 2.4k open PRs. I agree that it makes sense to audit changes, to check for potential breakage or bugs etc. But I don't think it helps (for newcomers and for keeping the queue under control) to leave nit-picky comments such as suggesting to write `callPackage foo/bar { }` rather than `callPackage foo/bar {}` or suggesting changes that are not related to the patch itself, but apply to (usually small) issues that happen to be in the same file.
> 
> Most of this can be automated by a tool and shouldn't require human effort. I personally think that focusing on things in changes that actually matter (well - from my PoV at least, feel free to disagree!) would make reviews more efficient and easier to justify (in contrast to "unreviewed" changes).

I 100% agree with nitpicking being the demotivator, but I'd like to keep that discussion as part of documenting the contributing guidelines, not here to prevent catch-all.

**Comment by Ma27** [comment]
> I'm not a fan of future-proofing by imposing requirements that might materialize one day. It increases the scope of work to be done. Maybe we'll stick with GitHub for another 10 years and then use something completely different, so a vendor-free solution is also useless at that point. Who knows?

Yeah I have no idea about the current state of the discussion, so let's leave that out here :+1:  

> I 100% agree with nitpicking being the demotivator, but I'd like to keep that discussion as part of documenting the contributing guidelines, not here to prevent catch-all.

Works for me. Wasn't sure if I'm the only one, so I thought it makes sense to mention it here (or in other words: I'm personally not 100% convinced at the moment that reviews bring an actual benefit in *every* case).

>Even though I mentioned that this conflicts with e.g. Domen's suggestion, I'm not sure I agree here:
>
> When PRs are supposed to be backported, we have a needs: port to stable-label. Or in other words: (at least) I usually bring this topic up when a PR is open. So finding out whether it's fine to backport is already part of the "main" PR and that's why I actually consider a backport (i.e. a push to a release branch) part of the merge process.
>
> Of course this doesn't apply to patches that are e.g. pretty huge or have non-trivial conflicts after a cherry-pick. Then a PR is obviously necessary.

Any comment on this? :) Because this may be relevant for the policy itself :)

**Comment by domenkozar** [comment]
> Any comment on this? :) Because this may be relevant for the policy itself :)

I imagine the new workflow to be applying that label opens up a new PR for a stable release. So then CI can kick in immediately and reviewing can be done.

Does that make sense? There was no question so I'm not sure this conflicts with my suggestion (now that I've clarified that I'm for PRs also for backports, just need to find good automation).

**Comment by Ma27** [comment]
> I imagine the new workflow to be applying that label opens up a new PR for a stable release. So then CI can kick in immediately and reviewing can be done.
>
> Does that make sense? There was no question so I'm not sure this conflicts with my suggestion (now that I've clarified that I'm for PRs also for backports, just need to find good automation).

Yes it does, that seems like a great idea to me!

Not sure if I'm just confused currently, but to me the OP reads like we *currently* have to file PRs for every single backport (as we don't have such a mechanism at the moment) which is something I don't really agree with for the reasons I stated above.

**Comment by jonringer** [comment]
I support this absolutely.

Even as a release manager, there's very few instances where I should be pushing directly (`git tag` and creating a release branch). Otherwise all "work" should really follow a PR process. Even when doing release notes, it was nice for others to find some mistakes I may have overlooked. No one is perfect, pull requests are a great to "improve" the quality of nixpkgs. If we had a proportional number of members with commit access, I would even go so far to say that you wouldn't be able to merge your own PR. However, this is a mostly volunteer organization, and people's free time is precious.

Also, nixpkgs conventions are subject to change over time as we develop better practices for packaging using nix. Some conventions which were acceptable long ago may not be now, as with https://github.com/NixOS/nixpkgs/pull/109461.

**Comment by domenkozar** [comment]
I've automated the process with GHA at https://github.com/NixOS/nixpkgs/pull/119859

**Comment by alyssais** [comment]
> We're **strongly discouraging direct pushes**

I'm not opposed to this, but I'm curious who "we" is here, and how this decision was arrived at.  It sounds very authoritative for a change whose RFC was closed.

**Comment by domenkozar** [comment]
> I'm not opposed to this, but I'm curious who "we" is here, and how this decision was arrived at. It sounds very authoritative for a change whose RFC was closed.

RFCs are not really mandatory for improving the process as per https://github.com/NixOS/rfcs#when-this-process-is-followed

As for "we": this has been discussed a lot of times and at the volume of nixpkgs committers we're not at 100% agreement (which isn't even possible) and there's no other agreed way to reach a consensus currently.

This issue is open for anyone to comment if they disagree with the practice, but we're kind of stuck: there's no process to accept change of process while the current workflow requires major changes.

I'd prefer we avoid perfectionism by only adding a warning to drive communication here for those that did push directly. Note that they are still allowed to push.

If there's wording that sounds less authoritative, feel free to suggest it :)

**Comment by alyssais** [comment]
> RFCs are not really mandatory for improving the process as per https://github.com/NixOS/rfcs#when-this-process-is-followed

Yeah, and I think that's really important. :)

> As for "we": this has been discussed a lot of times and at the volume of nixpkgs committers we're not at 100% agreement (which isn't even possible) and there's no other agreed way to reach a consensus currently.
>
> This issue is open for anyone to comment if they disagree with the practice, but we're kind of stuck: there's no process to accept change of process while the current workflow requires major changes.
>
> I'd prefer we avoid perfectionism by only adding a warning to drive communication here for those that did push directly. Note that they are still allowed to push.

Entirely reasonable and good.

> If there's wording that sounds less authoritative, feel free to suggest it :)

I might have said "It seems like we're close to a consensus that direct pushes to master are strongly discouraged".  I think that way, you'd getting across that most people agree, but without sounding like you're speaking on behalf of anybody else.

Thanks for taking this forward, anyway.  I think this is a good direction. (You may recall I was opposed to it some time ago, but I've overcome my objections -- GitHub's new CLI making the PR flow a lot nicer has helped a lot, as well as talking to @grahamc about it.)

**Comment by domenkozar** [comment]
Thanks @alyssais  :)

**Comment by Ma27** [comment]
@domenkozar regarding https://github.com/NixOS/nixpkgs/blob/master/.github/workflows/direct-push.yml I got a question:

I interpreted your suggestion to use some kind of bot for backporting (well, except for complex backports & conflicting revs) as a "mid-term" solution and that it's fine backport by a push (in most of the cases). To [quote myself](https://github.com/NixOS/nixpkgs/issues/118661#issuecomment-816764675):

> [...] finding out whether it's fine to backport is already part of the "main" PR and that's why I actually consider a backport (i.e. a push to a release branch) part of the merge process.

I skimmed through the comments after that and didn't find any disagreement. Now I'm a bit confused since the action states that pushes to release branches are discouraged as well (the action doesn't seem to be triggered currently btw - probably because it's not on release-20.09). Would you mind clarifying whether that's true per se (if so, I'd like to mention that I'm still against it - until we have a bot for that) or if it's just the wording that I misunderstood?

In the latter is the case, I'm happy to file a PR :)

**Comment by domenkozar** [comment]
Good question! My plan is to add backporting automation and then add the direct push action to the release branch as well.

I'm open to alternatives though :)

**Comment by Ma27** [comment]
> Good question! My plan is to add backporting automation and then add the direct push action to the release branch as well.

OK, I fully agree with that. I have a follow-up question then: do you plan to remove the action after the 21.05 branchoff from the new release branch if no automation is available until then (IIRC it's not so far to branch-off).

And, does it make sense then to file a PR where I alter the wording to make it clear that backports via a push are fine *until* we have the necessary automation?

**Comment by vcunat** [comment]
> IIRC it's not so far to branch-off

[May 21st](https://discourse.nixos.org/t/21-05-release-schedule/12528) is still a few weeks in future.

**Comment by domenkozar** [comment]
> OK, I fully agree with that. I have a follow-up question then: do you plan to remove the action after the 21.05 branchoff from the new release branch if no automation is available until then (IIRC it's not so far to branch-off).

I hope to get that going soon, but yeah. https://discourse.nixos.org/t/automating-pr-backports/12565 was quite so I'll prepare a prototype this week.

> And, does it make sense then to file a PR where I alter the wording to make it clear that backports via a push are fine until we have the necessary automation?

Adding some documentation how to use github cli to open a PR would be helpful for backport even for those backports that won't be trivial. I hope to tackle the backporting action soon so I'd prefer not to make changes in the opposite way :)

**Comment by Ma27** [comment]
> I hope to get that going soon, but yeah. https://discourse.nixos.org/t/automating-pr-backports/12565 was quite so I'll prepare a prototype this week.

Oh I wasn't sure how fast you'd plan to be :D 
But in that case, leaving it as-is is fine IMHO :)

**Comment by sumnerevans** [comment]
@Ma27 said https://github.com/NixOS/nixpkgs/issues/118661#issuecomment-816764675
> Also I'd like to suggest to reflect our way of reviewing things:
> 
> We currently have 2.4k open PRs. I agree that it makes sense to audit changes, to check for potential breakage or bugs etc. But I don't think it helps (for newcomers and for keeping the queue under control) to leave nit-picky comments such as suggesting to write `callPackage foo/bar { }` rather than `callPackage foo/bar {}` or suggesting changes that are not related to the patch itself, but apply to (usually small) issues that happen to be in the same file.
> 
> Most of this can be automated by a tool and shouldn't require human effort. I personally think that focusing on things in changes that actually matter (well - from my PoV at least, feel free to disagree!) would make reviews more efficient and easier to justify (in contrast to "unreviewed" changes).

Is there an existing PR/issue regarding adding some sort of linting system to the CI to help avoid nitpicky manual reviews? The home manager project has standardized on using `nixfmt` for example.

**Comment by jonringer** [comment]
> this year there's more freezing

For clarification for others, the freezing only applies to breaking changes (e.g. major version bumps) and PRs targeting staging for two weeks. Any non-breaking bump or PR targeting master will be unaffected.

However, there should be much less need to spam backport PRs, and ZHF will be on master for the first two weeks

**Comment by asymmetric** [comment]
It would be better to remove the link to this PR from the bot’s message, otherwise I think we will be forced to unsubscribe, lest we receive a notification every time someone pushes to master. Or is that part of the goal? ;)

Could it point to a text file instead?

**Comment by orivej** [comment]
I would not have known about this issue without the bot's comment. (I am subscribed to rfcs and have seen https://github.com/NixOS/rfcs/pull/79 . I thought it did not pass.)

**Comment by domenkozar** [comment]
> It would be better to remove the link to this PR from the bot’s message, otherwise I think we will be forced to unsubscribe, lest we receive a notification every time someone pushes to master. Or is that part of the goal? ;)
> 
> Could it point to a text file instead?

The goal is to have a public and transparent log of all direct pushes.

**Comment by Ma27** [comment]
> Is there an existing PR/issue regarding adding some sort of linting system to the CI to help avoid nitpicky manual reviews?

IIRC this was brought up a few times on IRC, but I don't know of any existing issue. However:

* IMHO it doesn't matter anyways whether we do e.g. `callPackage ./. {}` vs `callPackage ./. { }`.
* The last time this was discussed in `#nixos-dev` (IIRC) it was agreed that if we establish a formatter, it should be primarily used on new packages. Apart from the fact that formatting makes backports usually harder (because you'll run into way more merge conflicts), my experience is that most formatters (to be fair, I have to take a way closer look at both `nixfmt` and `nixpkgs-fmt`, but from what I've seen we'll have the problem here as well) make it harder to `git blame` a file. Despite the IMHO problematic name I generally use this to understand why a line of code looks the way it does when I don't understand it entirely.
  So what I want to say is, if we start using a formatter, I personally think that we should be careful with applying it to existing code.

**Comment by domenkozar** [comment]
I've moved the discussion about formatting to https://github.com/NixOS/nixpkgs/issues/120832

**Comment by vcunat** [comment]
We're getting some false positives, at least it seems that way to me (e.g. the last two above).  Anyone knows why?

**Comment by grahamc** [comment]
There is a race condition on GitHub's end that can cause these false positives. I'm not sure there is a 100% confidence way to do this?

**Comment by asymmetric** [comment]
@vcunat looks like in both cases, the "base branch" (which IIUC is the "target branch") was changed.

**Comment by vcunat** [comment]
Ah, OK.  Changing the target branch shouldn't happen too often.  We'll see.

**Comment by primeos** [comment]
I don't think changing the "base branch" is the problem here. Both https://github.com/NixOS/nixpkgs/commit/dfcfe50920c81302be069533514a16d4a48c6715 and https://github.com/NixOS/nixpkgs/commit/9468b073267c9159d79b89686fabd09367f74ac3 where pushed directly vs. using GH's merge button (in that case the merge commit would be signed by GitHub's OpenPGP key with ID 4AEE18F83AFDEB23). So the "you pushed a commit directly to master/release branch" part is actually technically correct but of course these kind of direct pushes are not lacking a PR so they could be considered fine (the problem is that our GitHub action doesn't seem to be able to verify that fast enough - one potential solution would be to delay the test for a few seconds if we want to allow manual merging/pushing of PRs via Git instead of GitHub).

I think the race condition is that GH takes a while to mark the corresponding PR as merged when pushing directly. Now the workflow shouldn't complain anymore:
```console
$ export GITHUB_REPOSITORY=NixOS/nixpkgs
$ export GITHUB_SHA=dfcfe50920c81302be069533514a16d4a48c6715
$ curl -s -H 'Accept: application/vnd.github.groot-preview+json' https://api.github.com/repos/$GITHUB_REPOSITORY/commits/$GITHUB_SHA/pulls | jq -r '.[] | select(.merge_commit_sha == "'$GITHUB_SHA'") | any'
true
```

**Comment by domenkozar** [comment]
Could someone merge a few PRs manually and time how long it takes for GitHub to propagate that?

Delaying the test seems fine to me, as long as it's predictably short for the vast majority of cases.

**Comment by vcunat** [comment]
_If_ it's the same as auto-closing the merged PR, I've been observing that delay as a couple seconds typically.

**Comment by 7c6f434c** [comment]
>Could someone merge a few PRs manually and time how long it takes for GitHub to propagate that?
>
>Delaying the test seems fine to me, as long as it's predictably short for the vast majority of cases.

Presumably one could do a check immediately, and only wait if it is a suspected direct push. Can we also ask this action to be always performed last?

**Comment by domenkozar** [comment]
Backporting action: https://github.com/NixOS/nixpkgs/pull/124273

**Comment by domenkozar** [comment]
You can now add `backport release-21.05` label to PRs and once they are mered, the bot will make a backport PR!

Note that it doesn't yet work for [rebase/squashed merges](https://github.com/zeebe-io/backport-action/issues/46), so be careful when merging :)

**Comment by 7c6f434c** [comment]
Hmm, could we do a retry after 30 seconds on author+committer commits, also on «Merge pull request #» commits?

Presumably checking that «cherry picked from» have the same diff is harder.

[Also, how come GitHub does not think I am participating in this issue and how to change its mind so that the notifications get sorted properly…]

**Comment by c0bw3b** [comment]
Yep I got scolded twice too.
It was squash-merges but the PR number was in the commit message.

**Comment by ryantm** [comment]
https://github.com/NixOS/nixpkgs/commit/24a9d6b163629148cbb546f4c0c0204ee7bde6bf#commitcomment-72106341 is a false positive because I clicked the merge button on https://github.com/NixOS/nixpkgs/pull/170041

**Comment by domenkozar** [comment]
Anyone has an idea why are we getting these false positives now? I'm really low on time :(

**Comment by 06kellyjac** [comment]
So the `ISMERGE` jq query does return "true" for the latest false positive

I imagine there is some kind of issue in querying the commit too quickly? Maybe we could put a healthy sleep on it, I don't think there's any need to rush this particular action since it only runs on master and release

**Comment by vcunat** [comment]
I don't use the button (in order to sign the merges), so my cases are a bit different, though they also are all just merging the current head of a PR.    Well, a couple times it happened that I merged a little older head, as `refs/pull/*/head` on GitHub lags a bit sometimes.

**Comment by doronbehar** [comment]
Why did I get a comment from the bot for https://github.com/NixOS/nixpkgs/commit/f521088c65bb4df4ebe686206660a9c6c4f8fb60 ? I merged it via the regular merge button...

**Comment by 06kellyjac** [comment]
> Why did I get a comment from the bot for [f521088](https://github.com/NixOS/nixpkgs/commit/f521088c65bb4df4ebe686206660a9c6c4f8fb60) ? I merged it via the regular merge button...

https://github.com/NixOS/nixpkgs/issues/118661#issuecomment-1109905716

**Comment by 7c6f434c** [comment]
Should the timeouts be bumped? Should the action be recognised as beyond repair and switched off?

**Comment by infinisil** [comment]
I just opened a new and updated RFC to disallow direct pushes: https://github.com/NixOS/rfcs/pull/156

**Comment by mweinelt** [comment]
> I wonder why I got [bdddb46#commitcomment-122875121](https://github.com/NixOS/nixpkgs/commit/bdddb46f4b058465ad53134d4a118671f9956662#commitcomment-122875121) on PR merge, it was not a direct push. What is wrong?

Because GitHub is eventually-consistent and at the time the action checked it did not yet attach the PR information to the API response.

**Comment by 7c6f434c** [comment]
> Because GitHub is eventually-consistent

Citation needed that it will ever become consistent. Some web merges are in the false positives in analytics run _months_ afterwards. 

GitHub just messes up metadata forever in a small enough fraction of cases that it never reaches the top of bugfix priority, and Nixpkgs large enough for this small fraction to translate to a noticeable stream of false positives.

**Comment by infinisil** [comment]
Direct pushes are now blocked since https://github.com/NixOS/nixpkgs/issues/249117, and the bot was removed in https://github.com/NixOS/nixpkgs/pull/249151, so we can officially close this issue now! :rocket:
